txnstats.go

```diff
@@ -0,0 +1,235 @@
+package main
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"net"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/juju/utils/parallel"
+	errgo "gopkg.in/errgo.v1"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+type Stats struct {
+	Collections map[string]CollectionStats
+	InProgress  InProgressStats
+	Log         LogStats
+}
+
+type CollectionStats struct {
+	DocCount    int `json:",omitempty"`
+	MaxQueued   int `json:",omitempty"`
+	MinQueued   int `json:",omitempty"`
+	TotalQueued int `json:",omitempty"`
+	// TODO count bad tokens too?
+}
+
+type LogStats struct {
+	DocCount int `json:",omitempty"`
+}
+
+type InProgressStats struct {
+	States    map[state]int
+	MaxOps    int `json:",omitempty"`
+	TotalOps  int `json:",omitempty"`
+	TotalTxns int `json:",omitempty"`
+}
+
+func main() {
+	args := commandLine()
+	session, err := dial(args)
+	if err != nil {
+		log.Fatal("cannot dial mongodb: %v", err)
+	}
+	db := session.DB("juju")
+	collNames, err := db.CollectionNames()
+	if err != nil {
+		log.Fatal("cannot dial mongodb: %v", err)
+	}
+	sort.Strings(collNames)
+	collStats := make([]CollectionStats, len(collNames))
+	run := parallel.NewRun(10)
+	for i, collName := range collNames {
+		if !wantCollectionStats(collName) {
+			continue
+		}
+		i, collName := i, collName
+		run.Do(func() error {
+			c := db.C(collName)
+
+			stats, err := getCollectionStats(c)
+			if err != nil {
+				return errgo.Notef(err, "cannot gather stats on %s: %v", c.Name, err)
+			}
+			collStats[i] = stats
+			return nil
+		})
+	}
+	stats := &Stats{
+		Collections: make(map[string]CollectionStats),
+	}
+	run.Do(func() error {
```

Comment:
If you're going to the effort for 'run.Do' you probably also want to Session.Copy() so that you can have the database queries also executing in parallel.